### PR TITLE
Added rough parametric plots

### LINKED ISSUE #128: Add support for Rough Parametric plots
## Details:
Hours Spent: 14

## Background
In the previous version of the application, the user could add parametric plots that have a separate expression for both `x` and `y`.

## Overview
Add support for parametric plots. In this first step only add the rough version and leave `TODO:` as a place holder for the robust plot computation.

## Acceptance Criteria
- [x] The user can input an `x(t)` expression that is parsed with respect to `t`
- [x] The user can input an `y(t)` expression that is parsed with respect to `t`
- [x] The user can input a range for the variable `t`

## Closing Notes

Overview of changes can be found in PR #129

### PR DESCRIPTION
**Pull request by RelativeForce**
# Issues
closes #128 

# Summary of changes
- Added `VariableName` as a parameter to the `checkExpression` function in `ExpressionInputController` so that it could be changed depending on which variable the expression would be in terms of.
- Renamed `Components.ExpressionInput` to `Components.ExpressionInput.FunctionExpressionInput` as it is now the component for simple functions plot. As part of this I also added `Function` as a prefix to all of the data constructors of `ExpressionInputMessage` (now called `FunctionExpressionInputMessage`) so they could be easily distinguished from the parametric versions.
- Moved `Status` and `DrawingStatus` to `ViewModels.Expression.Common` as they are used by both and all types of plots.
- Added `DrawingCommands` as a wrapper type for the `rough` and `robust` commands (along with there current `status`).
- Added `Components.ExpressionInput.ParametricExpressionInput` which handles all the input fields for the parametric plots.
- Renamed `ExpressionPlot` to `FunctionViewModel` and moved all the logic associated with drawing and updating its state into `ViewModels.Expression.Function` from `Components.Main.Action`.
- Added `ParametricViewModel` to hold all of the view data for a parametric plot. This includes the `xExpression` and `yExpression` and the `domain` (value range for `t`). It and its associated logic for drawing can be found in `ViewModels.Expression.Parametric`.
- Added `ExpressionViewModel` as the generic interface for any type of expression plot. Currently, it wraps `ParametricViewModel` and `FunctionViewModel` but in the future, it could be updated to include any type of plot (ie. fractals). The helper functions are included in the `ViewModels.Expression.Helper` module. All the generic functions for drawing the plots are included within the `ViewModels.Expression.Draw` module. These functions simply delegate to their appropriate functions.
- Obviously, there are some functions that are specific to each plot (such as updating their unique fields) so I added the `ViewModels.Expression.Unsafe`  module to house all of the functions that only work for one type of plot. These should only be used when 100% sure you are working with the right kind of plot.
- Removed `Components.ExpressionManager.Types` as it is redundant
- Refactored all the helper functions to do with `ExpressionPlot` from `Components.Main.Helper` into functions that are used by the `ExpressionViewModel`.
- Added capital `V` to the footer for the version as it starts the footer sentence
- Added `expectToMaybe` as a helper to convert an `Expect a` into a `Maybe a` where the `Left error` it converted to `Nothing`.
- Added `ExpectAff a` as an alis for `Aff (Either Error a)` as a tidy up. This also allowed me to add helper functions like `bindTo` for handling errors.
- Added `RoughParametricPlot` as a plot command
- Removed unused parameters from the `Plot.JobBatcher` functions
- Removed label string form the plot commands as it is not used
- Added `Plot.RoughParametricPlot` for calculating the rough parametric plots
- Added `ViewModels.Expression.Job` for all the job queue related functions for `ExpressionViewModel`

# Tests
N/A

### CODE REVIEW

File: src/Components/ExpressionManager/ExpressionManager.purs [hunk 1]
```diff
@@ -252,9 +271,15 @@ addPlotTab =
             [ className "form-inline" ]
             [ HH.button
                 [ className "btn btn-success btn-sm"
-                , onClickActionEvent Add
+                , onClickActionEvent AddParametric
                 ]
-                [ HH.text "+" ]
+                [ HH.text "P" ],
```
Comment (michalkonecny) [inline src/Components/ExpressionManager/ExpressionManager.purs]: I think "F" and "P" are a bit too short.  How about "+Fun" and "+Param"?

File: src/Misc/ExpectAff.purs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+bindTo :: forall a b. ExpectAff a -> (a -> ExpectAff b) -> ExpectAff b
+bindTo operation onSuccess = do
+  resultOrError <- operation
+  case resultOrError of
+    Left error -> pure $ Left error
+    Right result -> onSuccess result
+
+pureRight :: forall a. a -> ExpectAff a
+pureRight = pure <<< Right
```
Comment (michalkonecny) [inline src/Misc/ExpectAff.purs]: Run purty

File: src/Plot/RobustPlot.purs [hunk 3]
```diff
@@ -27,29 +27,21 @@ shouldLogSubsegments = false
...
+  evaluateWithX x = expectToMaybe $ evaluateDerivative variableMap expression
     where
     variableMap = [ Tuple "x" { value: x, derivative: one } ]
 
-    value = case evaluateDerivative variableMap expression of
-      Left _ -> Nothing
-      Right v -> Just v
-
-  evaluateWithX2 x = value
+  evaluateWithX2 x = expectToMaybe $ evaluateDerivative2 variableMap expression
```
Comment (michalkonecny) [inline src/Plot/RobustPlot.purs:41]: Nice simplification

File: src/ViewModels/Expresison/Common.purs [hunk 4]
```diff
@@ -0,0 +1,37 @@
+module ViewModels.Expression.Common where
```
Comment (michalkonecny) [inline src/ViewModels/Expresison/Common.purs]: Please correct the folder name.

File: src/ViewModels/Expresison/Draw.purs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+import Plot.JobBatcher (JobQueue, hasJobs)
+import Plot.Label (LabelledDrawCommand, drawRoughLabels)
+import Types (Size, XYBounds, Position)
+import ViewModels.Expression (ExpressionViewModel(..))
+import ViewModels.Expression.Common (AccuracyCalculator, DrawingStatus(..), Status(..), DrawingCommands, fromPixelAccuracy)
+import ViewModels.Expression.Function.Draw (drawRobustOnlyFunction, drawRoughAndRobustFunction, drawRoughOnlyFunction, overwiteFunctionAccuracy)
+import ViewModels.Expression.Helper (drawingStatus)
+import ViewModels.Expression.Parametric.Draw (drawRobustOnlyParametric, drawRoughAndRobustParametric, drawRoughOnlyParametric, overwiteParametricAccuracy)
+
+overwiteAccuracy :: Number -> AccuracyCalculator -> Int -> XYBounds -> ExpressionViewModel -> ExpectAff ExpressionViewModel
```
Comment (michalkonecny) [inline src/ViewModels/Expresison/Draw.purs]: Please correct the spelling of "overwrite".  This mistake appears in several names.

File: src/ViewModels/Expresison/Draw.purs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+drawRoughOnly size bounds (Parametric vm) =
+  mapExpectAff Parametric
+    $ drawRoughOnlyParametric
+        (fromPixelAccuracy size bounds)
+        size
+        bounds
+        vm
+
+allRobustComplete :: Array ExpressionViewModel -> Boolean
+allRobustComplete = all $ \vm -> DrawnRobust == drawingStatus vm
```
Comment (michalkonecny) [inline src/ViewModels/Expresison/Draw.purs]: Should we check only those models that have status Robust?  (I noticed that the status indicator reports that some robust plots are not completed even when their status is Rough.  I am not sure if it is happening due to this function or something else.)

**Review by michalkonecny — APPROVED** [review]
Very nice, only a few minor corrections to fix before merging. Thank you.

File: src/ViewModels/Expresison/Common.purs (hunk 4 above)
Comment (RelativeForce) [reply src/ViewModels/Expresison/Common.purs]: Fixed in latest commit

File: src/Misc/ExpectAff.purs (hunk 2 above)
Comment (RelativeForce) [reply src/Misc/ExpectAff.purs]: Fixed in latest commit

File: src/ViewModels/Expresison/Draw.purs (hunk 5 above)
Comment (RelativeForce) [reply src/ViewModels/Expresison/Draw.purs]: Fixed in latest commit (That's kind of embarrassing 😞)

File: src/ViewModels/Expresison/Draw.purs (hunk 6 above)
Comment (RelativeForce) [reply src/ViewModels/Expresison/Draw.purs]: It is and should be resolved by the latest commit

File: src/Components/ExpressionManager/ExpressionManager.purs (hunk 1 above)
Comment (RelativeForce) [reply src/Components/ExpressionManager/ExpressionManager.purs]: Updated in latest commit

**Pull request merged**